drivers/adc/adc_ti_am335x.c

```diff
@@ -0,0 +1,405 @@
+#define TI_ADC_STEPDELAY_SAMPLEDELAY_MAX (0xFFU)         /* Conversion Delay - Max */
+
+/* FIFO Threshold Register */
+#define TI_ADC_FIFO_THRESHOLD (40)
+
+/* FIFO Data Register */
+#define TI_ADC_FIFODATA_ADCDATA GENMASK(11, 0) /* FIFO Data Mask */
+
+#define DEV_CFG(dev)  ((const struct ti_adc_cfg *)(dev)->config)
+#define DEV_DATA(dev) ((struct ti_adc_data *)(dev)->data)
+#define DEV_REGS(dev) ((struct ti_adc_regs *)DEVICE_MMIO_GET(dev))
+
+struct ti_adc_cfg {
+	DEVICE_MMIO_ROM;
+	void (*irq_func)(const struct device *dev);
+	uint32_t open_delay[TI_ADC_TOTAL_CHANNELS];
+	uint8_t oversampling[TI_ADC_TOTAL_CHANNELS];
+	uint8_t fifo;
+};
+
+struct ti_adc_data {
+	DEVICE_MMIO_RAM;
+	struct adc_context ctx;
+	const struct device *dev;
+	uint8_t steps[TI_ADC_TOTAL_CHANNELS];
+	uint32_t fifo_irq_mask;
+	volatile const uint32_t *fifo_wc_ptr;
+	volatile const uint32_t *fifo_data_ptr;
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+	uint8_t chan_count;
+	uint8_t step_count;
+};
+
+/* Forward Declaration */
+static int ti_adc_sequencer_start(const struct device *dev);
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct ti_adc_data *data = CONTAINER_OF(ctx, struct ti_adc_data, ctx);
+	struct ti_adc_regs *regs = DEV_REGS(data->dev);
+
+	regs->CONTROL &= ~TI_ADC_CONTROL_ENABLE;
+
+	/* enable steps */
+	for (int chan = 0; chan < TI_ADC_TOTAL_CHANNELS; chan++) {
+		if (IS_BIT_SET(ctx->sequence.channels, chan)) {
+			regs->STEPENABLE |= BIT(data->steps[chan] + 1);
+		}
+	}
+
+	if (ti_adc_sequencer_start(data->dev) < 0) {
+		LOG_ERR("Sequencer failed to start");
+	};
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx, bool repeat)
+{
+	struct ti_adc_data *data = CONTAINER_OF(ctx, struct ti_adc_data, ctx);
+
+	if (repeat) {
+		data->buffer = data->repeat_buffer;
+	} else {
+		data->buffer += data->chan_count;
+	}
+}
+
+static int ti_adc_sequencer_start(const struct device *dev)
+{
+	struct ti_adc_data *data = DEV_DATA(dev);
+	struct ti_adc_regs *regs = DEV_REGS(dev);
+	uint64_t timeout = k_uptime_get();
+	uint32_t seq_status = regs->SEQ_STATUS;
+
+	while (FIELD_GET(TI_ADC_SEQ_STATUS_FSM, seq_status) != TI_ADC_SEQ_STATUS_FSM_IDLE &&
+	       FIELD_GET(TI_ADC_SEQ_STATUS_STEP, seq_status) != TI_ADC_SEQ_STATUS_STEP_IDLE) {
+		/* Timeout */
+		if (k_uptime_get() - timeout < (TI_ADC_IDLE_TIMEOUT_MS * data->chan_count)) {
+			return -ETIMEDOUT;
+		}
+
+		k_sleep(K_USEC(10)); /* sleep for 10us */
+
+		seq_status = regs->SEQ_STATUS;
+	}
+
+	regs->CONTROL |= TI_ADC_CONTROL_ENABLE;
+
+	return 0;
+}
+
+static int ti_adc_channel_setup(const struct device *dev, const struct adc_channel_cfg *chan_cfg)
+{
+	const struct ti_adc_cfg *cfg = DEV_CFG(dev);
+	struct ti_adc_data *data = DEV_DATA(dev);
+	struct ti_adc_regs *regs = DEV_REGS(dev);
+	const uint8_t chan = chan_cfg->channel_id;
+
+	if (chan >= TI_ADC_TOTAL_CHANNELS) {
+		LOG_ERR("Channel %d invalid, must be less than %d", chan, TI_ADC_TOTAL_CHANNELS);
+		return -EINVAL;
```

Comment:
This function could return `ENOTSUP` instead of `EINVAL` since those values are technically not supported but valid.

Reply:
thats fair

Reply:
On second thought, the function is supposed to only return `-EINVAL` in case of such values I think

https://github.com/zephyrproject-rtos/zephyr/blob/76e1fc7713a4a3f2b50c497afddec0364a34c10b/include/zephyr/drivers/adc.h#L712-L723

Reply:
oh, interesting. Quite some driver actually do use `ENOTSUP` ... but `-EINVAL` should do it too.